File: predict-mnist-rsample.jl
# random sample ensemble. accuracy: 0.97
using DelimitedFiles
using Random

square(x) = x * x

function sum_sq_err(xs::Vector, ys::Vector, start::Int)
    s = 0
    for i in start:length(xs)
        s += square(xs[i] - ys[i])
    end
    return s
end

function sum_sq_err_sample(xs::Vector, ys::Vector, start::Int, sample_indexes::Vector{Int})
    s = 0
    for i in sample_indexes
        s += square(xs[i] - ys[i])
    end
    return s
end

function predict(train_data::Vector{Vector{Float64}}, test::Vector{Float64}, nensemble::Int, nsample::Int)
    votes = Dict{Float64, Int}()
    for sample_num in 1:nensemble
        predicted = 0.0
        min_error = -1.0
        sample_indexes = shuffle(2:length(test))[1:nsample]
        
        for train in train_data
            error = sum_sq_err_sample(test, train, 1, sample_indexes)
            if error < min_error || min_error < 0
                predicted = train[1]
                min_error = error
            end
        end
        votes[predicted] = get(votes, predicted, 0) + 1
    end
    # find max vote
    maxk = 0.0
    maxv = 0
    for (k, v) in votes
        if v > maxv
            maxk = k
            maxv = v
        end
    end
    return maxk
end

function main()
    file_name_train = "mnist_train.csv"
    file_name_test = "mnist_test.csv"

    # train
    train_data = open(file_name_train) do io
        lines = readlines(io)
        [parse.(Float64, split(line, ",")) for line in lines]
    end
    println("train data loaded. rows: $(length(train_data))")

    # test
    n_rows = 0
    n_correct = 0
    open(file_name_test) do io
        for line in eachline(io)
            test = parse.(Float64, split(line, ","))
            n_rows += 1
            predicted = predict(train_data, test, 10, 600)
            answer = test[1]
            if predicted == answer
                n_correct += 1
            end
            accuracy = n_correct / n_rows
            println("row: $n_rows, predicted: $predicted, answer: $answer, accuracy: $accuracy")
        end
    end
end

main()

File: predict-mnist-rsample.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

func square(x float64) float64 {
	return x * x
}

func sumSqErr(xs []float64, ys []float64, start int) float64 {
	s := 0.0
	for i := start; i < len(xs); i++ {
		s += square(xs[i] - ys[i])
	}
	return s
}

func sumSqErrSample(xs []float64, ys []float64, start int, sampleIndexes []int) float64 {
	s := 0.0
	for _, i := range sampleIndexes {
		s += square(xs[i] - ys[i])
	}
	return s
}

func predict(trainData [][]float64, test []float64, nensemble int, nsample int) float64 {
	votes := make(map[float64]int)

	for sampleNum := 0; sampleNum < nensemble; sampleNum++ {
		predicted := 0.0
		minError := -1.0

		// Create a shuffled slice of indices from 1 to len(test)-1
		indices := rand.Perm(len(test) - 1)
		sampleIndexes := make([]int, nsample)
		for i := 0; i < nsample; i++ {
			sampleIndexes[i] = indices[i] + 1 // Adjust index to be 1-based
		}

		for _, train := range trainData {
			error := sumSqErrSample(test, train, 1, sampleIndexes)
			if error < minError || minError < 0 {
				predicted = train[0]
				minError = error
			}
		}
		votes[predicted]++
	}

	// Find max vote
	maxk := 0.0
	maxv := 0
	for k, v := range votes {
		if v > maxv {
			maxk = k
			maxv = v
		}
	}
	return maxk
}

func main() {
	fileNameTrain := "mnist_train.csv"
	fileNameTest := "mnist_test.csv"

	// train
	trainFile, err := os.Open(fileNameTrain)
	if err != nil {
		fmt.Println("Error opening train file:", err)
		return
	}
	defer trainFile.Close()

	var trainData [][]float64
	scanner := bufio.NewScanner(trainFile)
	for scanner.Scan() {
		line := scanner.Text()
		fields := strings.Split(line, ",")
		var row []float64
		for _, field := range fields {
			val, err := strconv.ParseFloat(field, 64)
			if err != nil {
				fmt.Println("Error parsing float in train data:", err)
				return
			}
			row = append(row, val)
		}
		trainData = append(trainData, row)
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading train file:", err)
		return
	}
	fmt.Printf("train data loaded. rows: %d\n", len(trainData))

	// test
	testFile, err := os.Open(fileNameTest)
	if err != nil {
		fmt.Println("Error opening test file:", err)
		return
	}
	defer testFile.Close()

	nRows := 0
	nCorrect := 0
	testScanner := bufio.NewScanner(testFile)
	for testScanner.Scan() {
		if err := testScanner.Err(); err != nil {
			fmt.Println("Error reading test file:", err)
			return
		}
		line := testScanner.Text()
		fields := strings.Split(line, ",")
		var test []float64
		for _, field := range fields {
			val, err := strconv.ParseFloat(field, 64)
			if err != nil {
				fmt.Println("Error parsing float in test data:", err)
				return
			}
			test = append(test, val)
		}
		nRows++
		predicted := predict(trainData, test, 10, 600)
		answer := test[0]
		if predicted == answer {
			nCorrect++
		}
		accuracy := float64(nCorrect) / float64(nRows)
		fmt.Printf("row: %d, predicted: %.1f, answer: %.1f, accuracy: %.4f\n", nRows, predicted, answer, accuracy)
	}
}
